src/rules/no_dupe_keys.rs

```diff
@@ -102,60 +212,111 @@ mod tests {
+    assert_lint_ok::<NoDupeKeys>(r#"var x = { '': 1, bar: 2 };"#);
+    assert_lint_ok::<NoDupeKeys>(r#"var x = { '': 1, ' ': 2 };"#);
+    assert_lint_ok::<NoDupeKeys>(r#"var x = { '': 1, [null]: 2 };"#);
+    assert_lint_ok::<NoDupeKeys>(r#"var x = { '': 1, [a]: 2 };"#);
+    assert_lint_ok::<NoDupeKeys>(r#"var x = { [a]: 1, [a]: 2 };"#);
+    assert_lint_ok::<NoDupeKeys>(r#"+{ get a() { }, set a(b) { } };"#);
+    assert_lint_ok::<NoDupeKeys>(r#"var x = { a: b, [a]: b };"#);
+    assert_lint_ok::<NoDupeKeys>(r#"var x = { a: b, ...c }"#);
+    assert_lint_ok::<NoDupeKeys>(
+      r#"var x = { get a() {}, set a (value) {} };"#,
+    );
+    assert_lint_ok::<NoDupeKeys>(r#"var x = ({ null: 1, [/(?<zero>0)/]: 2 })"#);
+    assert_lint_ok::<NoDupeKeys>(r#"var {a, a} = obj"#);
+    assert_lint_ok::<NoDupeKeys>(r#"var x = { 012: 1, 12: 2 };"#);
+    assert_lint_ok::<NoDupeKeys>(r#"var x = { 1_0: 1, 1: 2 };"#);
+    // nested
+    assert_lint_ok::<NoDupeKeys>(
+      r#"
+let x = {
+  y: {
+    foo: 0,
+    bar: 1,
+  },
+};
+"#,
+    );
   }
 
   #[test]
-  fn it_fails_when_there_are_duplicate_keys() {
+  fn no_dupe_keys_invalid() {
     assert_lint_err::<NoDupeKeys>(
       r#"var foo = { bar: "baz", bar: "qux" };"#,
       10,
     );
-  }
-
-  #[test]
-  fn it_fails_when_there_are_multiple_duplicate_keys() {
     assert_lint_err_n::<NoDupeKeys>(
       r#"var foo = { bar: "baz", bar: "qux", quux: "boom", quux: "bang" };"#,
       vec![10, 10],
     );
-  }
-
-  #[test]
-  fn it_fails_when_there_are_duplicate_string_keys() {
     assert_lint_err::<NoDupeKeys>(
       r#"var foo = { bar: "baz", "bar": "qux" };"#,
       10,
     );
-  }
-
-  #[test]
-  fn it_fails_when_there_are_duplicate_numeric_keys() {
     assert_lint_err::<NoDupeKeys>(r#"var foo = { 1: "baz", 0x1: "qux" };"#, 10);
-  }
-
-  #[test]
-  fn it_fails_when_there_are_duplicate_getter_keys() {
     assert_lint_err::<NoDupeKeys>(
       r#"var foo = { bar: "baz", get bar() {} };"#,
       10,
     );
-  }
-
-  #[test]
-  fn it_fails_when_there_are_duplicate_setter_keys() {
     assert_lint_err::<NoDupeKeys>(
       r#"var foo = { bar: "baz", set bar() {} };"#,
       10,
     );
+    assert_lint_err::<NoDupeKeys>(r#"var x = { a: b, ['a']: b };"#, 8);
+    assert_lint_err::<NoDupeKeys>(r#"var x = { '': 1, '': 2 };"#, 8);
+    assert_lint_err::<NoDupeKeys>(r#"var x = { '': 1, [``]: 2 };"#, 8);
+    assert_lint_err::<NoDupeKeys>(r#"var x = { 012: 1, 10: 2 };"#, 8);
+    assert_lint_err::<NoDupeKeys>(r#"var x = { 0b1: 1, 1: 2 };"#, 8);
+    assert_lint_err::<NoDupeKeys>(r#"var x = { 0o1: 1, 1: 2 };"#, 8);
+    // TODO(magurotuna): this leads to panic due to swc error
+    // It seems like tsc v4.0.2 cannot handle this either
+    // playground: https://www.typescriptlang.org/play?target=99&ts=4.0.2#code/MYewdgzgLgBCBGArGBeGBvAUDGBGMAXDACwBMANJgL4DcQA
+    // assert_lint_err::<NoDupeKeys>(r#"var x = { 1n: 1, 1: 2 };"#, 8);
```

Comment:
@magurotuna could you report that in SWC repo?

Reply:
@bartlomieju I thought I would do so, but as I commented, the TypeScript compiler can't parse this either. So I felt like it's okay for now. But I'll report it anyway 😄

Reply:
https://github.com/swc-project/swc/issues/1173

Filed an issue about it